i6_models/config.py

```diff
@@ -51,3 +53,22 @@ def _validate_types(self) -> None:
 
     def __post_init__(self) -> None:
         self._validate_types()
+
+
+ConfigType = TypeVar("ConfigType", bound=ModelConfiguration)
+ModuleType = TypeVar("ModuleType", bound=nn.Module)
+
+
+@dataclass
+class SubassemblyFactory(Generic[ConfigType, ModuleType]):
+    """
+    Dataclass for a combination of a Subassembly/Part and the corresponding configuration.
+    Also provides a function to construct the corresponding object through this dataclass
+    """
+
+    module_class: Callable[[ConfigType], ModuleType]
```

Comment:
I think this should be:
```suggestion
    module_class: Type[ModuleType]
```
You want to be able to have code like:
```python
assert isinstance(obj, opt.module_class)
```
So it would be good to have that guarantee that this `module_class` is really the class, and not any random callable.

Reply:
Btw, to make type hinting happy about this, you might need to adapt the `ModuleType` definition, do define it in a way that says that its `__init__` is in this way.

Reply:
@SimBe195

Reply:
> Btw, to make type hinting happy about this, you might need to adapt the ModuleType definition, do define it in a way that says that its __init__ is in this way.

And how would this work? Maybe by introducing a new protocol class? @albertz

Reply:
I haven't really tried it but I assume so.

Where exactly does it fail anyway? We could also just ignore the type checking for that line where it fails.

I don't think we should define a wrong type here just to make type checking happy. We should define the correct type, and fix the type checking, or disable type checking where it does not work.

Before disabling type checking for that line, I would first try to define the type in the correct way though.

Reply:
So I just tested in my PyCharm and `Callable[[ModelConfiguration], ModuleType]` causes 
```
class Test(nn.Module):
    def __init__(self, cfg: TestConfig):
        super().__init__()
        self.x = nn.Linear(cfg.param, cfg.param)
```
to not be recognized as matching type while `Type[ModuleType]`.
When constructing in both cases I did not get a typeguard error.

Reply:
I currently do not see a way to give typehints for how we expect the `__init__` of a class to look like. But maybe what we could do is at least to put an assertion into `__post_init__` that checks that the signature of `self.module_class.__init__` is correct.

Reply:
> and Callable[[ModelConfiguration], ModuleType] causes [...] to not be recognized as matching type

It should be `Callable[[ConfigType], ModuleType]`. But anyway I believe `Type[ModuleType]` + a runtime assertion is preferable.

Reply:
> So I just tested in my PyCharm and `Callable[[ModelConfiguration], ModuleType]` ...

But my suggestion was to actually not use this and use `Type[ModuleType]` instead? I'm a bit confused now.

> to not be recognized as matching type while `Type[ModuleType]`.

I think there is a typo? I'm not sure I understand the sentence correctly.

---

> It should be `Callable[[ConfigType], ModuleType]`.

No I think it really should be `Type[ModuleType]`. That is what we expect here.

> But anyway I believe `Type[ModuleType]` + a runtime assertion is preferable.

What runtime assertion do you mean? The code I posted was just for demonstration, which would require that it is `Type[ModuleType]`, not `Callable[[ConfigType], ModuleType]`. It was just an example to demonstrate why `Type[ModuleType]` is the correct thing here. 

---



> > Btw, to make type hinting happy about this, you might need to adapt the ModuleType definition, do define it in a way that says that its **init** is in this way.
> 
> And how would this work? Maybe by introducing a new protocol class? @albertz

Actually I tried this code now and I don't see any warning? As I asked before, where exactly do you get a warning when you use `Type[ModuleType]`? I don't see this. I tried with [this code](https://github.com/albertz/playground/blob/master/module_init_typing.py) and I don't see any warnings.

Reply:
> But my suggestion was to actually not use this and use Type[ModuleType] instead? I'm a bit confused now.

There are basically two approaches:

1. `module_class: Callable[[ConfigType], ModuleType]` which I suggested initially to type-hint the constructor correctly but this has the disadvantage that you can't use things like `isinstance(obj, opt.module_class)` like you said.
2. `module_class: Type[ModuleType]` which is more how we want it but gives no typehints for the constructor, so calling `self.module_class(self.cfg)` may raise a type warning. @Atticus1806  just wanted to test both options in PyCharm.

> No I think it really should be Type[ModuleType]. That is what we expect here.

Yes I agree but only if it's still checked that the constructor of the given `module_class` has the right signature. This way it will already raise an error in sisyphus when you create a `ModelConfiguration` and not only later in the training job.

> What runtime assertion do you mean?

An assertion in `__post_init__` that checks the signature of `module_class.__init__` as I wrote in my [previous message](https://github.com/rwth-i6/i6_models/pull/21#discussion_r1242084866).

> Actually I tried this code now and I don't see any warning? As I asked before, where exactly do you get a warning when you use Type[ModuleType]? I don't see this. I tried with [this code](https://github.com/albertz/playground/blob/master/module_init_typing.py) and I don't see any warnings.

In VSCode (Pylance) do get a warning:
![](https://github.com/rwth-i6/i6_models/assets/37951951/71d14f4c-c7d8-4784-bc30-c341626cb8ff)